client/my-sites/plans-features-main/hooks/use-action-callback.ts

```diff
@@ -0,0 +1,234 @@
+			);
+
+			if ( cartItemForStorageAddOn?.extra ) {
+				recordTracksEvent( 'calypso_signup_storage_add_on_upgrade_click', {
+					add_on_slug: cartItemForStorageAddOn.extra.feature_slug,
+				} );
+			}
+
+			if ( cartHandler ) {
+				cartHandler( cartItems );
+				return;
+			}
+
+			const planPath = cartItemForPlan?.product_slug
+				? getPlanPath( cartItemForPlan.product_slug )
+				: '';
+
+			const checkoutUrl = cartItemForStorageAddOn
+				? `/checkout/${ siteSlug }/${ planPath },${ cartItemForStorageAddOn.product_slug }:-q-${ cartItemForStorageAddOn.quantity }`
+				: `/checkout/${ siteSlug }/${ planPath }`;
+
+			const checkoutUrlWithArgs = addQueryArgs(
+				{ ...( withDiscount && { coupon: withDiscount } ) },
+				checkoutUrl
+			);
+
+			page( checkoutUrlWithArgs );
+			return;
+		},
+		[ siteSlug, withDiscount, planActionCallback, cartHandler ]
+	);
+
+	const processCartItemsForPlanAndAddOns = useCallback(
+		(
+			params: UseActionCallbackParams,
+			freeTrialPlanSlug: PlanSlug | undefined,
+			isFreeTrialCTA: boolean | undefined
+		) => {
+			const { cartItemForPlan, planSlug, selectedStorageAddOn } = params;
+
+			if ( isFreeTrialCTA ) {
+				processCartItems?.(
+					[ { product_slug: freeTrialPlanSlug as PlanSlug } ],
+					freeTrialPlanSlug
+				);
+				return;
+			}
+
+			const storageAddOnCartItem = selectedStorageAddOn &&
+				! selectedStorageAddOn.purchased && {
+					product_slug: selectedStorageAddOn.productSlug,
+					quantity: selectedStorageAddOn.quantity,
+					volume: 1,
+					extra: { feature_slug: selectedStorageAddOn?.featureSlugs?.[ 0 ] },
+				};
+
+			if ( cartItemForPlan ) {
+				processCartItems?.(
+					[ cartItemForPlan, ...( storageAddOnCartItem ? [ storageAddOnCartItem ] : [] ) ],
+					planSlug
+				);
+				return;
+			}
+
+			processCartItems?.( null, planSlug );
+		},
+		[ processCartItems ]
+	);
+
+	return useCallback(
+		( params: UseActionCallbackParams, freeTrialPlanSlug: PlanSlug | undefined ) => {
+			return ( isFreeTrialCTA?: boolean ) => {
+				const { planSlug } = params;
+				const upgradePlan = isFreeTrialCTA && freeTrialPlanSlug ? freeTrialPlanSlug : planSlug;
+
+				if ( ! isFreePlan( planSlug ) ) {
+					recordTracksEvent?.( 'calypso_plan_features_upgrade_click', {
+						current_plan: sitePlanSlug,
+						upgrading_to: upgradePlan,
+						saw_free_trial_offer: !! freeTrialPlanSlug,
+					} );
+				} else {
+					recordTracksEvent( 'calypso_signup_free_plan_click' );
+				}
+				processCartItemsForPlanAndAddOns?.( params, freeTrialPlanSlug, isFreeTrialCTA );
+			};
+		},
+		[ sitePlanSlug, processCartItemsForPlanAndAddOns ]
+	);
+}
+
+function composeActionCallback(
+	callback: () => void,
+	planActionCallback?: ( planSlug: PlanSlug ) => boolean
+) {
+	return ( { planSlug }: Pick< UseActionCallbackParams, 'planSlug' > ) => {
+		const earlyReturn = planActionCallback?.( planSlug );
+		return earlyReturn ? () => {} : callback;
+	};
+}
```

Comment:
Likewise. I am not sure I see the need for this extra function, especially under this name. It simply checks if some other callback returns true/false and sends back the callback passed to it. 

I think we are also overdoing it on abstractions around the name "..callback". It's incredibly confusing to me :/

Reply:
https://github.com/Automattic/wp-calypso/pull/88694#discussion_r1559917358

Reply:
For this specific case, there's is [a noticeable amount of duplication]( https://github.com/Automattic/wp-calypso/pull/88694/files#diff-414f2908c4fdd943075d97bbf756af6a4cc2a5b0faf5fdd045108cb0cb4b4974R171-R173 ) without the function. For the time being, I just moved it closer to where it's used for readability https://github.com/Automattic/wp-calypso/pull/88694/commits/a87a1dd7916e38810a1ec4eadc9eb0b60cce9681d.

Giving this more thought, one other point of confusion I _think_ I see with the term `callback` is that, for example, `composeActionCallback` can be read as _either_ "compose action-callback" or "compose-action callback". Perhaps if we used the term Cta instead? `composeActionCta`?